arangod/Agency/Node.h

```diff
@@ -67,189 +80,66 @@ enum Operation {
+using NodePtr = std::shared_ptr<Node const>;
 
 /// Any node may either be a branch or a leaf.
 /// Any leaf either represents an array or an element (_isArray variable).
 /// Nodes are are always constructed as element and can become an array through
 /// assignment operator.
 /// toBuilder(Builder&) will create a _vecBuf, when needed as a means to
 /// optimization by avoiding to build it before necessary.
-class Node final {
-  /// @brief Access private methods
-  friend class Store;
-
+class Node : public std::enable_shared_from_this<Node> {
  public:
   /// @brief Slash-segmented path
   using PathType = std::vector<std::string>;
+  using PathTypeView = std::span<std::string const>;
 
   /// @brief Child nodes
-  using Children = std::unordered_map<std::string, std::shared_ptr<Node>>;
+  using Children = ::immer::map<std::string, NodePtr>;
+
+  using Array = ::immer::flex_vector<velocypack::String>;
+
+  using VariantType = std::variant<Children, Array, velocypack::String>;
 
   /// @brief Split strings by forward slashes, omitting empty strings,
   /// and ignoring multiple subsequent forward slashes
   static std::vector<std::string> split(std::string_view str);
 
-  /// @brief Construct with name
-  explicit Node(std::string const& name);
-
-  /// @brief Copy constructor
-  Node(Node const& other);
-
-  /// @brief Move constructor
-  Node(Node&& other) noexcept;
-
-  /// @brief Construct with name and introduce to tree under parent
-  Node(std::string const& name, Node* parent);
-
-  /// @brief Construct with name and introduce to tree under parent
-  Node(std::string const& name, Store* store);
-
   /// @brief Default dtor
-  ~Node();
-
-  /// @brief Get name
-  std::string const& name() const;
-
-  /// @brief Get full path
-  std::string uri() const;
-
-  /// @brief Assignment operator
-  Node& operator=(Node const& node);
-
-  /// @brief Move operator
-  Node& operator=(Node&& node) noexcept;
-
-  /// @brief Apply value slice to this node
-  Node& operator=(arangodb::velocypack::Slice const&);
+  ~Node() = default;
 
   /// @brief Check equality with slice
   bool operator==(arangodb::velocypack::Slice const&) const;
 
-  /// @brief Check equality with slice
-  bool operator!=(arangodb::velocypack::Slice const&) const;
-
-  /// @brief Type of this node (LEAF / NODE)
-  NodeType type() const;
-
-  /// @brief Get node specified by path vector
-  Node& getOrCreate(std::vector<std::string> const& pv);
-
   /// @brief Get node specified by path vector
-  Node const* get(std::vector<std::string> const& pv) const;
+  NodePtr get(PathTypeView pv) const;
+  NodePtr get(std::initializer_list<std::string> const& pv) const;
 
   /// @brief Get node specified by path
-  Node const* get(
-      std::shared_ptr<cluster::paths::Path const> const& path) const;
-
-  /// @brief Get root node
-  Node const& root() const;
-
-  /// @brief Get root node
-  Node& root();
+  NodePtr get(std::shared_ptr<cluster::paths::Path const> const& path) const;
```

Comment:
There is another `get` method further down, let's move it here and explain, that `get` with an argument moves down the tree and returns a `NodePtr` or `nullptr` if the path does not exist.